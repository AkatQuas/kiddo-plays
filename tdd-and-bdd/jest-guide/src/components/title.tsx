import React, { CSSProperties } from 'react';

interface TitleProps {
  type: 'large' | 'small';
  title: string;
}

const largeStyle: CSSProperties = {
  fontSize: '2em',
  color: 'red',
};

const smallStyle: CSSProperties = {
  fontSize: '0.5em',
  color: 'green',
};

const styleMapper: Record<'small' | 'large', CSSProperties> = {
  small: smallStyle,
  large: largeStyle,
};

export const Title = (props: TitleProps) => {
  const { title, type } = props;

  return <p style={styleMapper[type]}>{title}</p>;
};
