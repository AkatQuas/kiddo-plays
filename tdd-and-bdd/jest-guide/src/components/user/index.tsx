import { Button } from 'antd';
import React, { useCallback } from 'react';
import { useAppDispatch, useAppSelector } from 'store/root';
import { selectUser, selectUserStatus } from 'store/user/selectors';
import { fetchUserThunk } from 'store/user/thunks';

export const User = ({}) => {
  const dispatch = useAppDispatch();

  const user = useAppSelector(selectUser);
  const status = useAppSelector(selectUserStatus);

  const handleClick = useCallback(async () => {
    await dispatch(fetchUserThunk());
  }, [dispatch]);
  return (
    <div>
      <h2>User Info</h2>
      {status === 'loading' ? <p>Loading...</p> : null}
      {user.id ? (
        <div>
          <p>ID: {user.id}</p>
          <p>Name: {user.name}</p>
          <p>Age: {user.age}</p>
        </div>
      ) : (
        <p>No information</p>
      )}

      <Button onClick={handleClick} type="primary">
        Load User
      </Button>
    </div>
  );
};
