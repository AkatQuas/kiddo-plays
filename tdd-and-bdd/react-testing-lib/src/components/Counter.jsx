import { useCallback, useState } from 'react';

const Counter = ({ initialValue = 2 }) => {
  const [count, setCount] = useState(initialValue);
  const increment = useCallback(() => {
    setCount((x) => x + 1);
  }, []);
  return (
    <div>
      <p>current value: {count}</p>
      <button onClick={increment}>Increment</button>
    </div>
  );
};

export default Counter;
