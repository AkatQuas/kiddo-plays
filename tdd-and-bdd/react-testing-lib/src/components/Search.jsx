function Search({ value, onChange, children }) {
  return (
    <div>
      <label htmlFor="search">{children}</label>
      <input id="search" type="text" value={value} onChange={onChange} />
    </div>
  );
}

export default Search;
