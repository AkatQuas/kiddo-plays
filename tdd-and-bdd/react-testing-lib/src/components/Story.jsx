import axios from 'axios';
import React, { useCallback, useState } from 'react';

const URL = 'http://hn.algolia.com/api/v1/search';

const Story = () => {
  const [stories, setStories] = useState([]);
  const [error, setError] = useState(null);

  const handleFetch = useCallback(async (e) => {
    let result;
    try {
      result = await axios.get(`${URL}?query=React`);
      setStories(result.data.hits);
    } catch (e) {
      setError(e);
    }
  }, []);
  return (
    <div>
      <button type="button" onClick={handleFetch}>
        Fetch Stories
      </button>
      {error ? <span>Something went wrong ...</span> : null}
      <ul>
        {stories.map((item) => (
          <li key={item.objectID}>
            <a href={item.URL}>{item.title}</a>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default Story;
