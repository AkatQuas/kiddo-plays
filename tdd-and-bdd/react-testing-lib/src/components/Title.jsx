function Title({ title = 'Hellow 42' }) {
  return <div>{title}</div>;
}

export default Title;
