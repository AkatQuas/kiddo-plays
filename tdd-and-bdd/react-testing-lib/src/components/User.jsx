import { useEffect, useState } from 'react';
const getUser = () => {
  return Promise.resolve({
    id: '1',
    name: 'Robinhood',
  });
};

const User = () => {
  const [user, setUser] = useState(null);
  useEffect(() => {
    getUser().then((u) => {
      setUser(u);
    });
  }, []);
  return user ? <p>Signed in as {user.name}</p> : null;
};

export default User;
