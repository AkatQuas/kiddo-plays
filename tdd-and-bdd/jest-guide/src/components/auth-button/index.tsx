import { Button, ButtonProps, message } from 'antd';
import { fetchUserRole, UserRole } from 'apis/user';
import classNames from 'classnames';
import React, { useEffect, useState } from 'react';
import styles from './styles.module.less';

type AuthButtonProps = ButtonProps;
const mapper = {
  user: 'User',
  admin: 'Admin',
};

export const AuthButton = ({
  children,
  className,
  ...rest
}: AuthButtonProps) => {
  const [userRole, setUserRole] = useState<UserRole>();

  useEffect(() => {
    fetchUserRole()
      .then((r) => {
        setUserRole(r.role);
      })
      .catch((e) => message.error(e.message));
    return () => {};
  }, []);

  return (
    <>
      <p>Role: {mapper[userRole!] ?? 'Unauthorized'}</p>
      <Button {...rest} className={classNames(className, styles.authButton)}>
        {children}
      </Button>
    </>
  );
};
