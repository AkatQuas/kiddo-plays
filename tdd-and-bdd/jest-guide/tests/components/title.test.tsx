import { render } from '@testing-library/react';
import { Title } from 'components/title';
import React from 'react';

describe('Title Component', () => {
  it('render large text', () => {
    const { baseElement } = render(<Title type="large" title="Large Text" />);
    expect(baseElement).toMatchSnapshot();
  });
  it('render small text', () => {
    const { baseElement } = render(<Title type="small" title="Small Text" />);
    expect(baseElement).toMatchSnapshot();
  });
});
