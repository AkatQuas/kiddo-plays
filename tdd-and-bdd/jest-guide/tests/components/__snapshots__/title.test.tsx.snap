// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title Component render large text 1`] = `
<body>
  <div>
    <p
      style="font-size: 2em; color: red;"
    >
      Large Text
    </p>
  </div>
</body>
`;

exports[`Title Component render small text 1`] = `
<body>
  <div>
    <p
      style="font-size: 0.5em; color: green;"
    >
      Small Text
    </p>
  </div>
</body>
`;
